libraries/AP_HAL/utility/RingBuffer.cpp

```diff
@@ -121,22 +119,42 @@ uint32_t ByteBuffer::peekbytes(uint8_t *data, uint32_t len)
         return 0;
     }
     uint32_t n;
-    const uint8_t *b = readptr(n);
+    auto b = readptr(n);
     if (n > len) {
         n = len;
     }
 
-    // perform first memcpy
-    memcpy(data, b, n);
-    data += n;
+    iovec[0].data = const_cast<uint8_t *>(b);
+    iovec[0].len = n;
 
     if (len > n) {
-        // possible second memcpy, must be from front of buffer
-        memcpy(data, buf, len-n);
+        iovec[1].data = buf;
+        iovec[1].len = len - n;
+
+        return 2;
     }
-    return len;
+
+    return 1;
+}
+
+/*
+  read len bytes without advancing the read pointer
+ */
+uint32_t ByteBuffer::peekbytes(uint8_t *data, uint32_t len)
```

Comment:
'peekbytes' is more elegant now, but slower and requires more stack memory, doesn't?